src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
+
+
+configure : Url -> Auth -> Result Error Config
+configure baseUrl auth =
+    Config baseUrl []
+        |> withAuthHeader auth
+        |> Debug.log "Kinto request"
+
+
+
+-- Dealing with answers from the Kinto server
+
+
+bodyDataDecoder : Json.Decoder Json.Value
+bodyDataDecoder =
+    ("data" := Json.value)
+
+
+extractData : String -> StatusCode -> StatusMsg -> Task Error Json.Value
+extractData body statusCode statusMsg =
+    let
+        responseResult =
+            Json.decodeString bodyDataDecoder body
+    in
+        case responseResult of
+            Ok data ->
+                Task.succeed data
+
+            Err _ ->
+                -- Is it an ErrorRecord json?
+                decodeError body statusCode statusMsg
+
+
+errorDecoder : Json.Decoder ErrorRecord
+errorDecoder =
+    Json.object4 ErrorRecord
+        ("errno" := Json.int)
+        ("message" := Json.string)
+        ("code" := Json.int)
+        ("error" := Json.string)
+
+
+decodeError : String -> StatusCode -> StatusMsg -> Task Error a
+decodeError body statusCode statusMsg =
+    case Json.decodeString errorDecoder body of
+        Ok errorRecord ->
+            Task.fail (KintoError statusCode statusMsg errorRecord)
+
+        Err msg ->
+            Task.fail (ServerError statusCode statusMsg body)
+
+
+toKintoResponse : Task Http.RawError Http.Response -> Task Error Json.Value
+toKintoResponse response =
+    (Task.mapError NetworkError response)
+        `Task.andThen` (handleResponse extractData)
+
+
+handleResponse :
+    (String -> StatusCode -> StatusMsg -> Task Error Json.Value)
+    -> Http.Response
+    -> Task Error Json.Value
+handleResponse handle response =
+    case response.value of
+        Http.Text body ->
+            handle body response.status response.statusText
+
+        _ ->
+            Task.fail
+                (ServerError
+                    response.status
+                    response.statusText
+                    "Response body is a blob"
+                )
+
+
+
+-- High level API
+
+
+performQuery : Result Error Config -> Endpoint -> Verb -> Task Error Json.Value
```

Comment:
It seems like the only way you can get a config error is by using `Basic` auth, because base64 can sometimes fail. (Wait, what? How can base64 encoding fail?? I just looked at the [source](https://github.com/truqu/elm-base64/blob/1.0.4/src/Base64.elm) and it seems like it shouldn't ever be able to fail.)

But assuming that it's possible for creating a `Config` to fail, I'd force the user to deal with it in one way or another. In Haskell, the "clean" thing to do is to "hide" the Config constructor and only export a "smart constructor", which is a function that can validate the input and return your `Result Error Config`. I think this is the way you're going for. The caller would then have to pattern match on the `Result` and decide what they wanted to do in the `Error` case (completely abort? prompt the user for another auth method?). But I don't think this can work quite as well in Elm since Config is just a type alias and a user can always create their own.

The "less clean and Ethan-approved way" is to throw an error using something like `Debug.crash` in the case where a config is "invalid", since that's something that shouldn't ever happen or at least happen very rarely. This lets you return a real `Config` in "all cases" by basically lying to the type checker. Serious Haskell People don't like this approach as much :)